### DISCUSSION TITLE
Understanding the corner cases for terraform version incompatibility

### DISCUSSION THREAD

**Discussion by yogesh-ti** [terraform-linters/tflint#1725] [Q&A] [answered]
My usecase:
I have a custom tflint ruleset plugin put in use with tflint. The ruleset only works upon resource and module blocks.
It tries to read a fixed tag name say `tag_name`
And based on some data either it provides linting on the `resource` keyword of a block or on some attribute inside the resource block

I have to make my ruleset work even at 0.13.x version of terraform and 1.x version.
In the [compatiblity docs](https://github.com/terraform-linters/tflint/blob/v0.22.0/docs/guides/compatibility.md) of older version there's one to one mapping of tflint version to terraform version for compatibility. And in the newer versions where the terraform packages are internalized instead of bundling the binary, support for 1.x versions is mentioned

I went through the release notes of terraform from 0.14.0 to 1.0.0 and didn't find any such syntax changes which would affect my use case.

Is it safe to use the latest tflint version of terraform code where the version is 0.13.0 or should I use multiple tflint version and maintain multiple plugins to make it work for different terraform versions (starting from 0.13.0) ?

**Comment by wata727** [comment, marked as answer]
My understanding is that there have been no major changes to the Terraform language since v0.12 (the transition to HCL2). In most cases, a configuration compatible with v0.13 will work with the latest TFLint version.

However, as Terraform officially mentions, the current Terraform language is only guaranteed to be compatible with v1.x.
https://developer.hashicorp.com/terraform/language/v1-compatibility-promises
For example, the `list`/`map` functions were removed in v0.15 so they work in v0.13 but not in the latest TFLint version.

For these reasons, it's necessary to maintain TFLint and plugins suitable for each version to ensure compatibility with any configuration. However, since v0.13 is no longer supported by TFLint, this approach is not recommended.

As a less proactive option, if there are no syntax differences that affect your use case, you can avoid maintaining multiple versions and just use the latest version of TFLint and plugins. However, it's ideal to upgrade to a supported version whenever possible.

**Reply by yogesh-ti** [reply]
Thanks for the clarification
